### PR TITLE
feat: 🎸 Add image widget

### PR DESCRIPTION
**Pull request by pavlo-vavruk**
Screenshot 2021-01-13 at 12 58 18
Add image widget

### CODE REVIEW

File: src/components/ImagePicker/ImagePicker.tsx [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+const ImagePicker: FC<{}> = () => {
+  const dispatch = useDispatch();
+  const { t } = useTranslation();
+
+  const [error, setError] = useState(null);
+  const [link, setLink] = useState('');
+
+  const isLink = (url: string) => {
+    return /(http(s?):)([/|.|\w|\s|-])*\.(?:jpg|gif|png)/.test(url);
```
Comment (apryka) [inline src/components/ImagePicker/ImagePicker.tsx:26]: are we considering `jpeg`, `svg` files?

File: src/components/ImageWidget/ImageWidget.tsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    getWidget(state, id)
+  );
+  const [placeholder, setPlaceholder] = useState({ width: 0, height: 0 });
+
+  useEffect(() => {
+    if (widgetRef.current) {
+      const { offsetWidth: width, offsetHeight: height } = widgetRef.current;
+      setPlaceholder({ width, height });
+    }
+  }, [widgetRef]);
```
Comment (apryka) [inline src/components/ImageWidget/ImageWidget.tsx:26]: please update dependency, should be `widgetRef.current`

File: src/modules/widgets/constants.ts [hunk 3]
```diff
@@ -9,6 +9,8 @@ export const SET_WIDGET_INITIALIZATION = '@widgets/SET_WIDGET_INITIALIZATION';
 export const INITIALIZE_WIDGET = '@widgets/INITIALIZE_WIDGET';
 export const INITIALIZE_CHART_WIDGET = '@widgets/INITIALIZE_CHART_WIDGET';
 export const EDIT_CHART_WIDGET = '@widgets/EDIT_CHART_WIDGET';
+export const EDIT_IMAGE_WIDGET = '@widgets/EDIT_IMAGE_WIDGET';
+export const IMAGE_WIDGET_CONFIGURE = '@widgets/IMAGE_WIDGET_CONFIGURE';
```
Comment (apryka) [inline src/modules/widgets/constants.ts:13]: please follow the naming convention, so the verb (action) should go first `CONFIGURE_IMAGE_WIDGET`
might be `SET_IMAGE_WIDGET` as it's shorter ;)

File: src/components/ImagePicker/ImagePicker.tsx (hunk 1 above)
Comment (maciejrybaniec) [reply src/components/ImagePicker/ImagePicker.tsx:26]: I believe this set of supported formats is enough for initial version. In my opinion this function should be move from `render` method as it's a pure function and do not need to be re-defined for each component re-render.

**Review by maciejrybaniec — CHANGES_REQUESTED** [review]

File: src/modules/widgets/saga.ts [hunk 4]
```diff
@@ -468,18 +494,44 @@ export function* editChartWidget({
...
+export function* editImageWidget({
+  payload,
+}: ReturnType<typeof editImageWidgetAction>) {
+  const { id } = payload;
+
+  const state = yield select();
+  const widgetId = getWidget(state, id).widget.id;
+
+  yield put(showImagePicker());
+  const action = yield take([SAVE_IMAGE]);
```
Comment (maciejrybaniec) [inline src/modules/widgets/saga.ts:506]: In my opinion - here we need to add a `HIDE_IMAGE_PICKER ` constant as user could also cancel the image edit flow.

**Review by maciejrybaniec — CHANGES_REQUESTED** [review]

**Review by maciejrybaniec — APPROVED** [review]

**Pull request merged**